avr-hal-generic/src/i2c.rs

```diff
@@ -425,51 +514,57 @@ macro_rules! impl_twi_i2c {
                 //    
                 //      .$twint().set_bit()
                 //
-                // But 'true' === 1 which in the language of the spec here means `cleared`...
+                // But in rust, a bool 'true' === 1. It's confusing that the datasheet says a bit
+                // is cleared by writing a logic one to it:
                 //
                 //    The TWINT Flag must be cleared by software by writing a logic one to it.
                 //
-                // Here in avr-device/src/generic.rs
-                //
-                //    impl<FI> FieldReader<bool, FI> {
-                //        /// Value of the field as raw bits.
-                //        #[inline(always)]
-                //        pub fn bit(&self) -> bool {
-                //            self.bits
-                //        }
-                //        /// Returns `true` if the bit is clear (0).
-                //        #[inline(always)]
-                //        pub fn bit_is_clear(&self) -> bool {
-                //            !self.bit()
-                //        }
-                //        /// Returns `true` if the bit is set (1).
-                //        #[inline(always)]
-                //        pub fn bit_is_set(&self) -> bool {
-                //            self.bit()
-                //        }
-                //    }
+                // And the FieldReader also explicitly uses 'clear' to refer to 0 and 'set' to
+                // refer to 1:
                 //
-                while self.p.$twcr.read().$twint().bit_is_clear() { }
-            }
+                //    impl<FI> FieldReader<bool, FI> { /// Value of the field as raw bits.
+                //    #[inline(always)] pub fn bit(&self) -> bool { self.bits } /// Returns `true`
+                //    if the bit is clear (0).  #[inline(always)] pub fn bit_is_clear(&self) ->
+                //    bool { !self.bit() } /// Returns `true` if the bit is set (1).
+                //    #[inline(always)] pub fn bit_is_set(&self) -> bool { self.bit() } }
+                //
+                // QUESTION: so is what I'm seeing a distinction between a _HARDWARE FLAG_ being 'cleared',
+                // (perhaps 'reset' is a better name?) vs. setting/clearing a _bit_ in a bitfield?
+                while self.p.$twcr.read().$twint().bit_is_clear() { } }
 
             fn transact(&mut self) {
+                // QUESTION: why are we setting TWEN here? It has already been set (above). Does it
+                // need to be set ? Why would it be unset? I looked through the datasheet and did
+                // not see any mention of this being turned off by hardware.
                 self.p.$twcr.write(|w| w.$twen().set_bit().$twint().set_bit());
+                // QUESTION: how do tight loops work on microcontrollers? Does this prevent the
+                // cpu from shutting down?
                 while self.p.$twcr.read().$twint().bit_is_clear() { }
```

Comment:
Do tight loops like this turn the CPU into a heater?

Reply:
Short answer: no

- First of all, this loop will only run for a very short time, not milliseconds or seconds.
- Microcontrollers generally support continuous execution of instructions for long periods of time because that's what they are usually meant to do.  There are some power-optimizations where you can put the MCU "to sleep" until an interrupt occurs but those are only used when you expect to get at least a few milliseconds of sleep time before the next interrupt.
- The reason you hear the fan spin up on a desktop machine when running such code for long periods of time is that your OS includes such optimizations and puts CPUs into idle-states when not in use.  Then, when someone is using 100% CPU, this no longer happens and the CPU runs continuously -> it heats up.  Even more, the OS might have a power governor that ramps up the clock speed when more CPU resources are used, leading to even more thermal.

In embedded/bare-metal, these tight loops waiting for hardware are very common.  They are by far the most efficient way to wait for a hardware even that is expected to arrive very soon.  They are even used to wait for software in some cases, when latency is critical (so called spin-locks).